I have a question

KVRTeJ opened the pull request.

KVRTeJ committed:
Создан класс BinaryTree и класс Node(узел). Добавлено обьявление методов

KVRTeJ committed:
Исправлена ошибка в методе isEmpty

KVRTeJ committed:
Добавлены новые определения методов

KVRTeJ committed:
комит

KVRTeJ committed:
Create QT .gitignore

KVRTeJ committed:
Добавлен конструктор для Node

KVRTeJ committed:
Изменен стандарт на gnu++23. Исправлена ошибка в объявлении дружественности BinaryTree Node.  Реализован конструктор по умолчанию. Реализованы методы add, toVector, printLeafs, printHorizontal

KVRTeJ committed:
Удаление ненужных файлов

KVRTeJ committed:
Добавлен метод find в двух вариациях.  Добавлена еще одна вариация метода toVector. Обработка исключения в Node::key()

KVRTeJ committed:
Начало реализации метода remove. Рефакторинг

KVRTeJ committed:
Рефакторинг

KVRTeJ committed:
Рефакторинг. Добавлен метод findParent

KVRTeJ committed:
Реализованы методы remove, clear, clearFrom, конструктор копирования, деструктор и оператор =

KVRTeJ committed:
Реализован метод copy. Рефакторинг remove и конструктора копирования(не работает)

KVRTeJ commented on Binary_Tree/Binary_Tree/Binary_Tree.cpp:
```diff
@@ -15,38 +15,7 @@ struct Data {
-        }
-    }
-
+    m_root = copy(other.m_root).m_root;
```
Он удаляет переменную типа BinaryTree, которую генерирует метод copy. У меня есть несколько вариантов: 
1) Сделать конструктор копирования почти идентичным методу copy(тогда этот кусок кода будет повторяться в 3 местах(оператор = ))
2) Сделать приватный метод copy(и публичный, где возвращается разыменованный результат приватного метода), который возвращает не сам объект, а указатель на него, но тогда придется выделять память под него в самом методе copy, но тогда мы не сможем его удалить, т.к. выделили память сами, а в деструкторе нельзя удалять сами объекты, или компилятор все же удалит эту память и утечки не будет?

tvz-041 replied:
Почему бы просто не использовать swap корней? В этом случае скопированное через copy дерево станет пустым, а this будет содержать требуемые узлы.

KVRTeJ replied:
Точно, спасибо!

KVRTeJ committed:
Реализован конструктор копирования, оператор =,  методы nodeCount, min, max, height(от корня объекта и произвольного),level(от корня объекта и произвольного),  balance(от корня объекта и произвольного)

KVRTeJ commented on Binary_Tree/Binary_Tree/Binary_Tree.hpp:
```diff
@@ -43,14 +48,21 @@ class BinaryTree {
     Node* findParent(Node* root, Node* child);
     
+    void compareBinaryTreeGreater(Node* root, int& buffer) const;
+    void compareBinaryTreeLess(Node* root, int& buffer) const;
```
Есть ли смысл сделать это через один метод, передавая в параметры анонимную функцию сравнения или передавая переменную типа bool(к примеру isOrderingLess) и сделать ветвление в самой функции?

tvz-041 replied:
Не совсем понимаю почему эти методы так называются, если по факту это:
```c++
int min(Node* root) const;
int max(Node* root) const;
```

KVRTeJ replied:
Сначала я хотел сделать это одним методом, и думал назвать compareBinaryTree, но мне показалось, что это того не стоит(была идея передавать анонимную функцию сравнения). Исправлено на min и max.

KVRTeJ commented:
Сделал все методы(без итераторов и GUI)

KVRTeJ committed:
Начата реализация итератора(не работает)

KVRTeJ committed:
Итератор теперь не шаблонный. Исправление ошибок

KVRTeJ committed:
Исправление ошибок

tvz-041 commented on Binary_Tree/Binary_Tree/Binary_Tree.hpp:
```diff
@@ -2,5 +2,150 @@
+    : m_key(key), m_left(left), m_right(right)
+    {}
+    
+    int key() const {return (this ? m_key : 0);}
```
На всякий случай напомню: проверка this на null не имеет смысла.

KVRTeJ replied:
Исправлено

KVRTeJ replied:
<img width="1149" alt="Снимок экрана 2024-03-15 в 00 46 44" src="https://github.com/KVRTeJ/Chuder/assets/144002117/31c0f339-ed8a-4143-8161-124342ab29a3">

Вызвал ->key() для nullptr

tvz-041 commented on Binary_Tree/Binary_Tree/Binary_Tree.hpp:
```diff
@@ -2,5 +2,150 @@
+    
+    ~TemplateIterator() = default;
+    
+    Node* operator * () {return *m_iter;}
```
Не хватает проверки на валидность итератора.

KVRTeJ replied:
Исправлено

KVRTeJ committed:
Исправвление ошибок

KVRTeJ committed:
Изменен метод TemplateIterator::operator * ()

KVRTeJ committed:
Полностью реализован итератор(константный и неконстантный)

Для этого пришлось объявить метод levelNodes шаблонным, т.к. перегрузить его нелья(отличается только возвращаемым значением, а если убрать модификатор const у метода, то это будет выглядеть не очень, да и в const методе нельзя вызывать non const метод). Определение вынесено в hpp файл.  Добавлен метод на проверку валидности итератора, обновлен операатор *

KVRTeJ committed:
Реализован метод вывода в консоль уровней, при помощи итератора

KVRTeJ commented:
Сделал все методы, но у меня вопрос по поводу remove(). При удалении корня заменять его ближайший лист?

KVRTeJ committed:
Revert "Revert "Создан проект""

This reverts commit cea274207f23018b0044f97425d04f6ffd88e2ef.

KVRTeJ committed:
Revert "Создан проект"

This reverts commit d9feee9a018eafdc1938e4880d109f637f362eb5.

KVRTeJ committed:
Создан проект

KVRTeJ committed:
Обновление файлов

tvz-041 commented on Binary_Tree/Binary_Tree/Binary_Tree.cpp:
```diff
@@ -1 +1,517 @@
+        return;
+    }
+    
+    std::vector<Node* > leafs = getLeafs(root);
+    while(leafs[0] != root) {
+        for(auto it = leafs.begin(); it != leafs.end(); ++it) {
+            remove(*it);
+        }
+        leafs = getLeafs(root);
+    }
```
Очень дорого для метода, который должен удалить все узлы, вызывать для каждого узла функцию, которая будет опять проходить от корня для поиска родителя.

KVRTeJ replied:
Хорошо, я хотел максимально уменьшить повторение кода, поэтому решил пойти именно этим способом. Добавил новый метод, clearFromInclusiveRoot, который рекурсией удаляет все узлы, включая root, и в методе clearFrom вызываю его для левого и правого потомков. Исправлено.

tvz-041 commented on Binary_Tree/Binary_Tree/Binary_Tree.cpp:
```diff
@@ -1 +1,517 @@
+    }
+}
+
+bool BinaryTree::balance() const {
+    return balance(m_root);
+}
+
+bool BinaryTree::balance(Node* root) const {
```
Всё же `isBalanced`, `balanced` было бы верно, если бы функция возвращала баланс узла.

KVRTeJ replied:
Исправлено

tvz-041 commented on Binary_Tree/Binary_Tree/Binary_Tree.cpp:
```diff
@@ -1 +1,517 @@
+            deleteChild(data);
+            
+        } else {
+            data.replacementNode = findParent(data.target, nullptr);
```
Не понимаю смысл этого вызова.

KVRTeJ replied:
Если вы про вызов findParent, то этот метод, когда передается nullptr, ищет ближайший лист

tvz-041 commented on Binary_Tree/Binary_Tree/Binary_Tree.cpp:
```diff
@@ -1 +1,517 @@
+    
+    data.nodeParent = findParent(m_root, data.target);
+    
+    auto deleteChild {
```
Возможно, лучше `finishRemove`. В структуре нет поля с именем `child`, также приставка finish намекает, что ниже мы будем делать что-то ещё перед этим действием (а именно, убирать потомков у `target`).

KVRTeJ replied:
Спасибо! Исправвлено

KVRTeJ committed:
Рефакторинг
